stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+    // directly. Other possible choices were to throw exceptions or to
+    // coninue formatting without representing these values correctly
+    // (like JavaScript's JSON.stringify).
+    if (isNaN(value)) {
+      Some(InvalidNumber("NaN is not allowed in JsonNumber"))
+    } else if (value < 0.0W) {
+      Some(InvalidNumber("-Infinity is not allowed in JsonNumber"))
+    } else {
+      Some(InvalidNumber("Infinity is not allowed in JsonNumber"))
+    }
+  }
+}
+
+@unsafe
+let printNumber = (value: Number, buffer: Buffer.Buffer) => {
+  from WasmI32 use { (&), (==), (!=), (<<), (>>) }
+
+  let ptr = WasmI32.fromGrain(value)
+  let ret = if ((ptr & 1n) != 0n) {
+    printNumberWasmI32(untagSimpleNumber(value), buffer)
+    None
+  } else if ((ptr & 7n) == Tags._GRAIN_GENERIC_HEAP_TAG_TYPE) {
+    let tag = WasmI32.load(ptr, 0n)
+    match (tag) {
+      t when t == Tags._GRAIN_BOXED_NUM_HEAP_TAG => {
+        let numberTag = WasmI32.load(ptr, 4n)
+        match (numberTag) {
+          t when t == Tags._GRAIN_INT64_BOXED_NUM_TAG => {
+            let asWasmI64 = WasmI64.load(ptr, _INT64_BOXED_VALUE_OFFSET)
+            printNumberWasmI64(asWasmI64, buffer)
+            None
+          },
+          t when t == Tags._GRAIN_BIGINT_BOXED_NUM_TAG => {
+            Buffer.addString(BI.bigIntToString10(ptr), buffer)
+            None
+          },
+          t when t == Tags._GRAIN_RATIONAL_BOXED_NUM_TAG => {
+            from WasmF64 use { (/) }
+            // As a compromnise, convert the rational into a float
+            // and then print that. This means there can be
+            // precision loss even for values that could be printed
+            // exactly in decimal notation, but doing otherwise
+            // could both compromise perfomance and confuse
+            // developers as most JSON libraries in other languages
+            // likely only deal with integers and floats.
+            // Note that the ECMA-404 standard only specifies how
+            // numbers should be formatted as text (in decimal
+            // notation), so we could go the extra mile and at
+            // least for some numbers try printing the exact
+            // number. For example those that have a power of 10
+            // in the denominator.
+            let asWasmF64 = WasmF64.convertI32S(
+              Numbers.boxedRationalNumerator(ptr)
+            ) /
+              WasmF64.convertI32S(Numbers.boxedRationalDenominator(ptr))
+
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          t when t == Tags._GRAIN_FLOAT64_BOXED_NUM_TAG => {
+            let asWasmF64 = WasmF64.load(ptr, _Float64_BOXED_VALUE_OFFSET)
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          _ => {
+            throw UnknownNumberTag
+          },
+        }
+      },
+      _ => {
+        throw UnknownNumberTag
+      },
+    }
+  } else {
+    throw UnknownNumberTag
+  }
+  ignore(value)
```

Comment:
@ospencer this is something for the GC? Can we add a comment because this is easily something that can be deleted in a cleanup/refactor.

Reply:
Yes. Any time `fromGrain` is used and the value isn't used later in the function, you need this `ignore` to keep the reference alive so it's not prematurely freed.